Core/automation/lib/python/core/metadata.py

```diff
@@ -0,0 +1,297 @@
+                item=item_name, namespace=name, value=str(value), configuration=str(configuration)))
+            self.save()
+    
+    def __iter__(self):
+        return iter(self._configuration)
+
+    def __len__(self):
+        return len(self._configuration)
+
+    def __setitem__(self, key, value):
+        return self.set_config_value(key, value)
+
+    def __getitem__(self, key):
+        return self.get_config_value(key)
+
+    def __delitem__(self, key):
+        return self.delete_config_value(key)
+
+    def load(self):
+        '''
+        Loads the namespace from the metadata registry.
+        THIS WILL OVERWRITE ANY UNSAVED CHANGES TO THIS INSTANCE!
+        '''
+        log.debug("Metadata: Getting metadata for item '{item}' from namespace '{namespace}'".format( \
+            item=self._item_name, namespace=self._name))
+        # read from registry
+        metadata = MetadataRegistry.get(self._key_uid)
+        log.debug("Metadata: {metadata}".format( \
+            metadata=str(metadata)))
+        # parse data
+        if metadata is not None: # namespace exists
+            self._value = _resolve_type(metadata.value)
+            # load all configuration items into namespace dict
+            self._configuration = {}
+            for key, value in metadata.configuration.iteritems():
+                self._configuration[str(key)] = _resolve_type(value)
+
+    def save(self):
+        '''
+        Saves the namespace to the metadata registry.
+        (This is done automatically when changes to the value or configuration are made
+        unless the "save" flag was set to false when making those changes)
+        '''
+        # convert all values to strings
+        if self._value is None:
+            strValue = "None"
+        else:
+            strValue = str(self._value)
+        strConfiguration = {}
+        for key, value in self._configuration:
+            strConfiguration[key] = str(value)
+        # save to the registry
+        MetadataRegistry.add(Metadata(self._key_uid, strValue, strConfiguration))
+        del strConfiguration
+
+    def remove(self):
+        '''
+        Deletes this namespace from the registry.
+        This instance will be preserved and can be written back to the registry.
+        '''
+        MetadataRegistry.remove(self._key_uid)
+
+    def set_value(self, value, save=True):
+        '''
+        Sets the namespace "value".
+        Set "save" flag to False to skip saving to openHAB. You must take care to manually save
+        using the "save()" method or setting another element with the "save" flag set to True.
+        '''
+        self._value = value
+        if save:
+            self.save()
+
+    @value.setter
+    def value(self, value):
+        '''
+        Sets the namespace "value" and saves to the registry.
+        '''
+        return self.set_value(value)
```

Comment:
Testing the first example (metadata_namespace), I get an error when using the module. Removing this function fixed it, but I did not test all methods yet. There is another function named `value` at L233, but I think this issue was with `value.setter`. The error was...

```
2019-05-12 10:59:48.624 [ERROR] [org.openhab.core.automation.module.script.internal.ScriptEngineManagerImpl] - Error during evaluation of script 'file:/opt/openhab2/conf/automation/jsr223/python/personal/test/testScript.py': NameError: name 'value' is not defined in <script> at line number 53
```

Reply:
I think this is one I added after I had done my tests. I think I know what it's complaining about though. When defining custom properties as methods like that you have to start with the getter then have the setter. If memory serves (I'm on my phone) it's like this:

```python
@property
def value(self):
    return self._value

@value.setter
def value(self, value):
    self._value = value
```

I may have forgotten to add that getter. I'll look at it tonight

Reply:
Oops, you said there is another at line 233, that's the issue, the getter needs to be first